README.md
# BAMSAUR

![build](https://github.com/bbartholdy/BAMSAUR/workflows/R/badge.svg)

This package provides functions for the age-at-death estimation of nonadult human skeletal remains based on dental wear. The method was developed using a Dutch post-Medieval reference population; however, the package also allows the inclusion of other (more appropriate) reference samples. The main function allows the age-at-death estimation of nonadults based on a single wear score calculated from all available teeth, and additional functions allow the user to upload and evaluate their own reference sample, from which to base the age-at-death estimations.
Users who are not familiar with the R-environment can also take advantage of a user-friendly graphical user interface (GUI).

https://websaur.shinyapps.io/bamsaurapp/

# Installation

1. To install BAMSAUR, R (>= 3.5.1) must first be installed (https://cran.r-project.org/).

```r
#2. Open R, and install the package "devtools" by entering the following into the R-console:
install.packages("devtools")

#3. The BAMSAUR package can then be downloaded and installed:
devtools::install_github("bbartholdy/BAMSAUR")
library(BAMSAUR)
```

# Usage

The BAMSAUR package uses the following R-packages:

```r
c("AICcmodavg", "earth", "ggfortify", "ggplot2", "shiny")
```

The method was developed on Middenbeemster, a 19th century Dutch skeletal collection containing individuals with documented age-at-death. The method uses a 9-stage dental wear scoring system, applicable to the entire deciduous and permanent (excluding third molars) dentitions. The average wear scores were used to develop a quadratic regression model from which to base age-at-death predictions. Age ranges are calculated using 68% prediction intervals (PIs). Users can also elect to use linear and cubic regression, or multivariate adaptive regression splines (MARS) for age-at-death predictions. Additional functions can be used to evaluate potential reference samples by developing linear, quadratic, and cubic regression models, as well as a MARS model (using the 'earth' package), and provides information on the developed models such as r-squared values, accuracies from leave-one-out cross validations (LOOCV), predictive residual error sum of squares (PRESS), Akaike and Bayesian information criterion (AIC and BIC), and precision (based on the average size of predicted age ranges). It also returns plots (using 'ggplot2') for each of the regression models.

# Functions

## BAMSAUR age-at-death estimation

`BAMSAUR(wear, data = NULL, pop = "MB11", rank = 2, interval = "prediction", level = 0.68, mars.int = TRUE, ...)`

`BAMSAUR.mars(wear, data = NULL, pop = "MB11", level = 0.68, varmod.method = "earth", nfold = n-1, ncross = 3)`

The main function which can be used to predict age-at-death based on dental wear by entering an obtained wear score, following the dental wear method outlined in the main article (Bartholdy et al., under review). The Middenbeemster data is built-in, so if pop = "MB11" is chosen, the data input is not required. If pop = "other", a data frame with a column for age and a second column for wear scores can be inserted. The function returns the input wear score, the age prediction, age range as +- years, and lower and upper intervals. The default method is a quadratic (rank = 2) model with 68%PIs, but "confidence" intervals and any level between 0 and 1 (0 - 100%), not inclusive, can be chosen. The input 'mars.int' allows the user to apply MARS-sized intervals to the estimates from simple regression models. The recommended setting for estimations based on the MB11 reference sample is TRUE, as this reference sample contains some individuals without documented age, and the MARS-sized age intervals can better account for this error. If 'mars.int = FALSE', regular confidence/prediction intervals are provided.

The BAMSAUR.mars function serves the same purpose as the BAMSAUR function, but uses a MARS model for the age-at-death estimation. Only "prediction" intervals are supported for MARS models. The default number of cross validations (ncross) is 3 to reduce computation time.

Examples:

```r
#The following example calculates the ages-at-death with MARS 68%PIs of 10 random wear scores
wear <- runif(10,0,16)
BAMSAUR(wear)
#Ages-at-death using a linear model and 95%CIs
BAMSAUR(wear, rank = 1, interval = "confidence", level = 0.95, mars.int = F)

#Age-at-death calculation using MARS
wear <- runif(10,0,16)
BAMSAUR.mars(wear)
#Age-at-death calculation using 90%PIs and a custom dataset
data <- MBsimple[1:25]
BAMSAUR.mars(wear, data = data, pop = "other", level = 0.9)
```

## BAMSAUR best-fit function

"It's not the sample size that matters, it's the goodness-of-fit!"

`BAMSAUR.bff(data, interval = "prediction", level = 0.68, varmod.method = "earth", nfold = n-1, ncross = 3, ...)`

This function can be used to evaluate a dataset to determine its feasibility as a reference sample for estimating age-at-death in the main BAMSAUR function. The data input must contain one column with known ages-at-death (whether osteologically estimated or documented age) and a second column with the associated wear scores. The function returns a summary of the developed linear, quadratic, and cubic regression models, as well as a MARS model. It also returns values to indicate goodness-of-fit of the various models, including r-squared values, accuracies from the LOOCV, PRESS, AIC (corrected for small sample sizes), BIC, and average age range (to indicate precision), as well as diagnostic plots.

Examples:

```r
#the following expression evaluates the sample with 68% prediction intervals as age ranges
MBex <- BAMSAUR.bff(MBsimple)
#the following code accesses the quadratic plot
MBex$quad.plot
```

## BAMSAUR shiny app

`runBAM()`

This launches the BAMSAUR shiny app, which has similar functionality as the individual functions.


Example:
```r
##Not run: runBAM()
```

## BAMSAUR data

The BAMSAUR data contains two datasets: MBsimple and MBdata. MBsimple contains a column with known ages-at-death and associated wear scores for the Middenbeemster sample, which is built-in to the BAMSAUR and BAMSAUR.mars functions.
MBdata contains all the wear scores for the individual teeth of each specimen that was used for the analysis, and is included for sharing and reproducibility purposes.

# References

Bartholdy, B.P., Hoogland, M.L.P, and Waters-Rist, A., under review. How Old Are You Now? A new ageing method for nonadults based on dental wear. International Journal of Osteoarchaeology. doi: 10.1002/oa.2758.

Hadley Wickham, Jim Hester and Winston Chang (2018). devtools: Tools to Make Developing R Packages Easier. R package version 1.13.6. https://CRAN.R-project.org/package=devtools

Hadley Wickham, Peter Danenberg and Manuel Eugster (2018). roxygen2: In-Line Documentation for R. R package version 6.1.0. https://CRAN.R-project.org/package=roxygen2

JJ Allaire, Jeffrey Horner, Vicent Marti and Natacha Porte (2017). markdown: 'Markdown' Rendering for R. R package version 0.8. https://CRAN.R-project.org/package=markdown

Marc J. Mazerolle (2017) AICcmodavg: Model selection and multimodel inference based on (Q)AIC(c). R package version 2.1-1. https://cran.r-project.org/package=AICcmodavg.

Milborrow, S. Derived from mda:mars by Trevor Hastie and Rob Tibshirani. Uses Alan Miller's Fortran utilities with Thomas Lumley's leaps wrapper. (2019). earth: Multivariate adaptive regression splines. R package version 4.7.0. https://CRAN.R-project.org/package=earth
publication

R Core Team (2018). R: A language and environment for statistical computing. R Foundation for Statistical Computing, Vienna, Austria. URL https://www.R-project.org/.

Wickham, H. (2015). R packages: organize, test, document, and share your code. O'Reilly Media, Inc.

Wickham, H. (2016). ggplot2: Elegant graphics for data analysis. New York: Springer-Verlag.

Winston Chang, Joe Cheng, JJ Allaire, Yihui Xie and Jonathan McPherson (2018). shiny: Web Application Framework for R. R package version 1.2.0. https://CRAN.R-project.org/package=shiny

Yuan Tang, Masaaki Horikoshi, and Wenxuan Li. "ggfortify: Unified Interface to Visualize Statistical Result of Popular R Packages." The R Journal 8.2 (2016): 478-489.


NEWS.md
# NEWS

Updates to the BAMSAUR package
## BAMSAUR_0.0.0.9002
### 13-06-2019

* Compatibility issues fixed

## BAMSAUR_0.0.0.9001
### 11-04-2019

* Erroneous code corrected in the BAMcv.lm function, which was providing incorrect accuracy rates.

## BAMSAUR_0.0.0.9000
### 11-03-2019

* Added option to apply MARS-sized prediction intervals to the simple regression models. This is recommeded when using the MB11 population, as the age range provided by the MARS model can better account for the error of the added undocumented-age individuals.

* Leave-one-out cross validation now performed on MARS models, allowing PRESS to be calculated.

* General housekeeping.


R/BAMSAURbff.R
#' Sample evaluation
#'
#' This function can be used to evaluate the viability of a sample for use as a reference population. It provides information on the accuracy of a sample using LOOCV, as well as details on the regression analyses performed on the sample.
#' @param data Data frame containing one column for ages-at-death and a second column for the associated wear scores.
#' @param interval Character. The type of age interval used. Can be either "prediction" or "confidence" intervals. Confidence intervals are not available for 'earth' class objects.
#' @param level Numeric. Determines level of confidence or prediction intervals. Can be any number between 0 and 1 (not inclusive), but 0.68, 0.90, or 0.95 is recommended. To run the function without intervals, enter 0.
#' @param ... Additional arguments can be passed to the 'earth' function. See ?earth for more details.
#' @details BAMSAUR.bff uses the "lm" and "earth" functions for the regression analyses. It automatically provides the accuracy of the sample being evaluated, based on whether the actual age of an individual falls within the estimated age range from linear and quadratic regression analyses (using the described LOOCV method).
#' It also provides the mean of the estimated age ranges, as an indicator of method precision, as well as the predicted residual error sum of squares (PRESS), the r-squared values, Akaike information criterion (AIC; AICc for sample sizes below 80), Bayes information criterion(BIC), and generalised cross validation (GCV).
#'
#' @return The function automatically returns a summary of the analysis, with the overall accuracy, the average age range (precision), the sum of squares of the predictive residuals (PRESS),
#' the R-squared value, AIC(c), and BIC.
#' It also includes a list containing the following items:
#'
#' \describe{
#'  \item{\code{lin.plot}}{a plot of the linear regression with prediction/confidence intervals}
#'
#'  \item{\code{quad.plot}}{a plot of the quadratic regression with prediction/confidence intervals}
#'
#'  \item{\code{cub.plot}}{a plot of the cubic regression with prediction/confidence intervals}
#'
#'  \item{\code{linear}}{class 'lm' object from the linear regression analysis.}
#'
#'  \item{\code{quadratic}}{class 'lm' object from the quadratic regression analysis.}
#'
#'  \item{\code{cubic}}{class 'lm' object from the cubic regression analysis.}
#'
#'  \item{\code{lin.data}}{the data from the linear LOOCV analysis.}
#'
#'  \item{\code{quad.data}}{the data from the quadratic LOOCV analysis.}
#'
#'  \item{\code{cub.data}}{the data from the quadratic LOOCV analysis.}
#'
#'  \item{\code{accuracy}}{A table with the percentage of estimates whose age ranges contain the actual age, the percentage of estimates within 1 year of the actual age, and within 2 years of the actual age.}
#' }
#' @note If the slope of the regression is not statistically different (2 standard deviations) from 0 or 1, a warning message will appear.
#' @seealso \code{\link[stats]{lm}}
#' @example inst/BAMSAURbffex.R
#' @import AICcmodavg ggplot2
#' @importFrom grDevices rgb
#' @importFrom stats lm na.omit AIC BIC rstandard predict
#' @importFrom graphics plot
#' @export BAMSAUR.bff
BAMSAUR.bff <- function(data, interval = "prediction", level = 0.68, ...) {
  cat("Calculating..."); cat("\n")
    data <- as.data.frame(data)
    data <- na.omit(data)
    age.data <- data[,1]
    wear.data <- data[,2]
    n <- as.numeric(length(age.data))

#cubic regression
    wear2.data <- wear.data^2
    wear3.data <- wear.data^3
    cub <- lm(age.data ~ wear.data + wear2.data + wear3.data)
    out.cub <- summary(cub)
    cub.rsq <- out.cub$r.squared
    cub.s <- out.cub$sigma
    cub.SSres <- round(sum(cub$residuals^2), digits = 2)
    cub.fitted <- cub$fitted.values
    cub.df <- cub$df.residual
    intercept <- cub$coefficients[[1]]
    a <- cub$coefficients[[2]]
    a2 <- cub$coefficients[[3]]
    a3 <- cub$coefficients[[4]]

  if(n/2 > 40){
      cub.aic <- AIC(cub)
    } else {
      cub.aic <- AICcmodavg::AICc(cub)
    }
    cub.bic <- BIC(cub)
    cub.PRESS <- sum(rstandard(cub, type = "predictive")^2)
    coef.sig <- a +- (2*out.cub$coefficients[[2,2]]) != 1|0
    coef2.sig <- a2 +- (2*out.cub$coefficients[[3,2]]) != 1|0
    coef3.sig <- a3 +- (2*out.cub$coefficients[[4,2]]) != 1|0

    if(coef.sig == "FALSE" | coef2.sig == "FALSE" | coef3.sig == "FALSE"){
      warning("slope of the cubic regression is not significantly different from zero")
    }

  #quadratic regression
  wear2.data <- wear.data^2
  quad <- lm(age.data ~ wear.data + wear2.data)
  out.quad <- summary(quad)
  quad.rsq <- out.quad$r.squared
  quad.s <- out.quad$sigma
  quad.SSres <- round(sum(quad$residuals^2), digits = 2)
  quad.fitted <- quad$fitted.values
  quad.df <- quad$df.residual
  intercept <- quad$coefficients[[1]]
  a <- quad$coefficients[[2]]
  a2 <- quad$coefficients[[3]]

  if(n/2 > 40){
  quad.aic <- AIC(quad)
  } else {
  quad.aic <- AICcmodavg::AICc(quad)
  }
  quad.bic <- BIC(quad)
  quad.PRESS <- sum(rstandard(quad, type = "predictive")^2)
  coef.sig <- a +- (2*out.quad$coefficients[[2,2]]) != 1|0
  coef2.sig <- a2 +- (2*out.quad$coefficients[[3,2]]) != 1|0

  if(coef.sig == "FALSE" | coef2.sig == "FALSE"){
    warning("slope of the quadratic regression is not significantly different from zero")
  }

  #linear regression
  lin <- lm(age.data ~ wear.data)
  out.lin <- summary(lin)
  lin.rsq <- out.lin$r.squared
  lin.s <- out.lin$sigma
  lin.SSres <- round(sum(lin$residuals^2), digits = 2)
  lin.fitted <- lin$fitted.values
  lin.df <- lin$df.residual
  intercept.lin <- lin$coefficients[[1]]
  a.lin <- lin$coefficients[[2]]

  if(n/2){
    lin.aic <- AIC(lin)
  } else {
    lin.aic <- AICcmodavg::AICc(lin)
  }
  lin.bic <- BIC(lin)
  lin.PRESS <- sum(rstandard(lin, type = "predictive")^2)
#is the slope of the regression significantly different from 0 or 1?
  coef.sig.lin <- a.lin +- (2*out.lin$coefficients[[2,2]]) != 1|0
    if(coef.sig == "FALSE"){
    warning("slope of the linear regression is not significantly different from zero")
  }

#LOOCV of the data
CV.cub <- BAMcv.lm(cub, interval = interval, level = level)
#CV.cub <- BAMSAUR.LOOCV(cub, interval=interval, level=level)

#CV.quad <- BAMSAUR.LOOCV(quad, interval=interval, level = level)
CV.quad <- BAMcv.lm(quad, interval = interval, level = level)

#CV.lin <- BAMSAUR.LOOCV(lin, interval = interval, level = level)
CV.lin <- BAMcv.lm(lin, interval = interval, level = level)

#Accuracy table output
acc.table <- as.data.frame(matrix(nrow = 3, ncol = 3), row.names = c("linear", "quadratic", "cubic"))
acc.table[1,1] <- CV.lin$accuracy
acc.table[1,2] <- CV.lin$accuracy.1
acc.table[1,3] <- CV.lin$accuracy.2
acc.table[2,1] <- CV.quad$accuracy
acc.table[2,2] <- CV.quad$accuracy.1
acc.table[2,3] <- CV.quad$accuracy.2
acc.table[3,1] <- CV.cub$accuracy
acc.table[3,2] <- CV.cub$accuracy.1
acc.table[3,3] <- CV.cub$accuracy.2
colnames(acc.table) <- c("total", "1yr", "2yrs")

#Creation of the plots
lin.plot <- BAM.plot(lin, interval=interval, level=level)
quad.plot <- BAM.plot(quad, interval=interval, level=level)
cub.plot <- BAM.plot(cub, interval=interval, level=level)

#Age ranges for calculation of precision
pred.cub <- suppressWarnings(predict(cub, interval = interval, level = level))
Int.cub <- (pred.cub[,3] - pred.cub[,2]) / 2
#Int.cub <- Interval(wear = wear.data, wear.data = wear.data, interval = interval, level = level, df = quad.df, s = quad.s)
  upp.int.cub <- cub.fitted + Int.cub
  low.int.cub <- cub.fitted - Int.cub

  pred.quad <- suppressWarnings(predict(quad, interval = interval, level = level))
  Int.quad <- (pred.quad[,3] - pred.quad[,2]) / 2
  #Int.quad <- Interval(wear = wear.data, wear.data = wear.data, interval = interval, level = level, df = quad.df, s = quad.s)
  upp.int.quad <- quad.fitted + Int.quad
  low.int.quad <- quad.fitted - Int.quad

  pred.lin <- suppressWarnings(predict(lin, interval = interval, level = level))
  Int.lin <- (pred.lin[,3] - pred.lin[,2])/2
  #Int.lin <- Interval(wear = wear.data, wear.data = wear.data, interval = interval, level = level, df = lin.df, s = lin.s)
  upp.int.lin <- lin.fitted + Int.lin
  low.int.lin <- lin.fitted - Int.lin
Int.lin <- round(Int.lin, 2)
lin.prec <- round(mean(Int.lin), 2)
Int.quad<- round(Int.quad, 2)
quad.prec <- round(mean(Int.quad), 2)
Int.cub <- round(Int.cub, 2)
cub.prec <- round(mean(Int.cub), 2)
#output
  cat("\n")
  cat(message("Linear model"));
  cat("R-squared:"); cat("\t"); cat(round(lin.rsq, digits = 2)); cat("\n")
  cat("PRESS:"); cat("\t"); cat("\t"); cat(round(lin.PRESS, digits = 2)); cat("\n")
  cat("AIC:"); cat("\t"); cat("\t"); cat(round(lin.aic, 2)); cat("\n")
  cat("BIC:"); cat("\t"); cat("\t"); cat(round(lin.bic, 2)); cat("\n")
  cat("Accuracy:"); cat("\t"); cat(round(CV.lin$accuracy, digits = 2)); cat("%"); cat("\n")
  cat("Average range:"); cat("\t"); cat("+-"); cat(" "); cat(lin.prec); cat(" "); cat("years"); cat("\n")
  cat("\t"); cat("(min, max: "); cat(min(Int.lin));cat(", "); cat(max(Int.lin)); cat(")"); cat("\n")
  cat("\n"); cat(message("Quadratic model"));
  cat("R-squared:"); cat("\t"); cat(round(quad.rsq, digits = 2)); cat("\n")
  cat("PRESS:"); cat("\t"); cat("\t"); cat(round(quad.PRESS, digits = 2)); cat("\n")
  cat("AIC:"); cat("\t"); cat("\t"); cat(round(quad.aic, 2)); cat("\n")
  cat("BIC:"); cat("\t"); cat("\t"); cat(round(quad.bic, 2)); cat("\n")
  cat("Accuracy:"); cat("\t"); cat(round(CV.quad$accuracy, digits = 2)); cat("%"); cat("\n")
  cat("Average range:"); cat("\t"); cat("+-"); cat(" "); cat(quad.prec); cat(" "); cat("years"); cat("\n")
  cat("\t"); cat("(min, max: "); cat(min(Int.quad));cat(", "); cat(max(Int.quad)); cat(")"); cat("\n")
  cat("\n"); cat(message("Cubic model"));
  cat("R-squared:"); cat("\t"); cat(round(cub.rsq, digits = 2)); cat("\n")
  cat("PRESS:"); cat("\t"); cat("\t"); cat(round(cub.PRESS, digits = 2)); cat("\n")
  cat("AIC:"); cat("\t"); cat("\t"); cat(round(cub.aic, 2)); cat("\n")
  cat("BIC:"); cat("\t"); cat("\t"); cat(round(cub.bic, 2)); cat("\n")
  cat("Accuracy:"); cat("\t"); cat(round(CV.cub$accuracy, digits = 2)); cat("%"); cat("\n")
  cat("Average range:"); cat("\t"); cat("+-"); cat(" "); cat(cub.prec); cat(" "); cat("years"); cat("\n")
  cat("\t"); cat("(min, max: "); cat(min(Int.cub));cat(", "); cat(max(Int.cub)); cat(")"); cat("\n")

  invisible(list("lin.plot" = lin.plot, "quad.plot" =  quad.plot, "cub.plot" = cub.plot, "linear" = lin, "quadratic" = quad,
                "cubic" = cub, "lin.data" = CV.lin$out, "quad.data" = CV.quad$out, "cub.data" = CV.cub$out, "accuracy" = acc.table))
  }
